source/common/network/apple_dns_impl.cc

```diff
@@ -0,0 +1,315 @@
+     * cancelling the operation, you can no longer wait for a callback *without* MoreComing set, to
+     * tell you [to flush] (the operation has been canceled, so there will be no more callbacks).
+     *
+     * [FURTHER] An implication of the collective
+     * kDNSServiceFlagsMoreComing flag for shared connections is that this
+     * guideline applies more broadly -- any time you cancel an operation on
+     * a shared connection, you should perform all deferred updates for all
+     * operations sharing that connection. This is because the MoreComing flag
+     * might have been referring to events coming for the operation you canceled,
+     * which will now not be coming because the operation has been canceled.
+     */
+    parent_.flushPendingQueries();
+  } else {
+    delete this;
+  }
+}
+
+void AppleDnsResolverImpl::PendingResolution::onDNSServiceGetAddrInfoReply(
+    DNSServiceFlags flags, uint32_t interface_index, DNSServiceErrorType error_code,
+    const char* hostname, const struct sockaddr* address, uint32_t ttl) {
+  ENVOY_LOG(warn,
+            "DNS for {} resolved with: flags={}[MoreComing={}, Add={}], interface_index={}, "
+            "error_code={}, hostname={}",
+            dns_name_, flags, flags & kDNSServiceFlagsMoreComing ? "yes" : "no",
+            flags & kDNSServiceFlagsAdd ? "yes" : "no", interface_index, error_code, hostname);
+  ASSERT(interface_index == 0);
+
+  if (error_code == kDNSServiceErr_DefunctConnection || error_code == kDNSServiceErr_Timeout ||
+      error_code == kDNSServiceErr_Refused) {
+    // Current query gets a failure status
+    if (!pending_cb_) {
+      ENVOY_LOG(warn, "[Error path] Adding to queries pending callback");
+      pending_cb_ = {ResolutionStatus::Failure, {}};
+      parent_.addPendingQuery(this);
+    } else {
+      ENVOY_LOG(warn, "[Error path] Changing status for query already pending flush");
+      pending_cb_->status_ = ResolutionStatus::Failure;
+    }
+
+    ENVOY_LOG(warn, "[Error path] DNS Resolver flushing queries pending callback");
+    parent_.flushPendingQueries();
+
+    // The main sd ref if destroyed here because a callback with an error is good indication that
+    // the connection to the DNS deamon is faulty and needs to be torn down.
+    //
+    // Deallocation of the MainSdRef __has__ to happend __after__ flushing queries. Flushing queries
+    // deallocates individual refs, so deallocating the main ref aheadd would cause deallocation of
+    // invalid individual refs per dns_sd.h
+    parent_.deallocateMainSdRef();
+    parent_.initializeMainSdRef();
+
+    return;
+  }
+
+  // At this point all known non-success cases have been dealt with. Assert that the error code is
+  // no error.
+  // POINT FOR DISCUSSION: Crash otherwise to alert consumers that there are unadressed edge cases?
+  // stat?
+  RELEASE_ASSERT(error_code == kDNSServiceErr_NoError,
```

Comment:
Seems fragile as new versions of the library might introduce new error codes. Serendipitously I was just updating https://github.com/envoyproxy/envoy/pull/13096,

Reply:
+1 I would not crash here and would handle as a generic error above.